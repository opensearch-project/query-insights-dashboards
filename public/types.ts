/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DataSourcePluginStart } from '../../../src/plugins/data_source/public';
import { DataSourceManagementPluginSetup } from '../../../src/plugins/data_source_management/public';
import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';

/* eslint-disable @typescript-eslint/no-empty-interface */
export interface QueryInsightsDashboardsPluginSetup {}
export interface QueryInsightsDashboardsPluginStart {}
export interface QueryInsightsDashboardsPluginStartDependencies {
  dataSource?: DataSourcePluginStart;
}
export interface QueryInsightsDashboardsPluginSetupDependencies {
  dataSourceManagement?: DataSourceManagementPluginSetup;
}
/* eslint-enable @typescript-eslint/no-empty-interface */

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
  dataSource?: DataSourcePluginStart;
}
