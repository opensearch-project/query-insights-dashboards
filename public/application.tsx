/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { HashRouter as Router } from 'react-router-dom';
import { AppMountParameters, CoreStart } from '../../../src/core/public';
import { QueryInsightsDashboardsApp } from './components/app';

export const renderApp = (core: CoreStart, { element }: AppMountParameters) => {
  ReactDOM.render(
    <Router>
      <QueryInsightsDashboardsApp core={core} />
    </Router>,
    element
  );

  return () => ReactDOM.unmountComponentAtNode(element);
};
