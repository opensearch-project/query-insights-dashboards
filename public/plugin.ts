/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  AppMountParameters,
  CoreSetup,
  CoreStart,
  DEFAULT_NAV_GROUPS,
  Plugin,
} from '../../../src/core/public';
import {
  QueryInsightsDashboardsPluginSetup,
  QueryInsightsDashboardsPluginSetupDependencies,
  QueryInsightsDashboardsPluginStart,
  QueryInsightsDashboardsPluginStartDependencies,
} from './types';
import { PLUGIN_NAME } from '../common';

export class QueryInsightsDashboardsPlugin
  implements
    Plugin<
      QueryInsightsDashboardsPluginSetup,
      QueryInsightsDashboardsPluginStart,
      {},
      QueryInsightsDashboardsPluginStartDependencies
    > {
  public setup(
    core: CoreSetup,
    deps: QueryInsightsDashboardsPluginSetupDependencies
  ): QueryInsightsDashboardsPluginSetup {
    // Register an application into the side navigation menu
    core.application.register({
      id: PLUGIN_NAME,
      title: 'Query Insights',
      // @ts-ignore
      description:
        'Monitor and analyze queries using the Query Insights Plugin, which ranks queries based on their resource utilization (CPU, JVM) and latencies',
      category: {
        id: 'opensearch',
        label: 'OpenSearch Plugins',
        order: 2000,
      },
      order: 5000,
      async mount(params: AppMountParameters) {
        // Load application bundle
        const { renderApp } = await import('./application');
        // Get start services as specified in opensearch_dashboards.json
        const [coreStart, depsStart] = await core.getStartServices();
        // Render the application
        return renderApp(
          coreStart,
          depsStart as QueryInsightsDashboardsPluginStartDependencies,
          params,
          deps.dataSourceManagement
        );
      },
    });

    // Registration for new navigation
    core.chrome.navGroup.addNavLinksToGroup(DEFAULT_NAV_GROUPS.dataAdministration, [
      {
        id: PLUGIN_NAME,
        category: {
          id: 'performance',
          label: 'Performance',
          order: 9000,
          euiIconType: 'managementApp',
        },
        order: 200,
      },
    ]);

    return {};
  }

  public start(_core: CoreStart): QueryInsightsDashboardsPluginStart {
    return {};
  }

  public stop() {}
}
