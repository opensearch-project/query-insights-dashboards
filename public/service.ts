/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

// public/services.ts
import type { CoreStart } from '../../../src/core/public';

let coreStartServices: CoreStart;

export function setStartServices(coreStart: CoreStart) {
  coreStartServices = coreStart;
}

export function getSavedObjectsClient() {
  return coreStartServices.savedObjects.client;
}

export function getRouteService() {
  return coreStartServices.http;
}
