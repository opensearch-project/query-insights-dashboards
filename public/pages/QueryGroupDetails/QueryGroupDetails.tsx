/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AppMountParameters, CoreStart } from 'opensearch-dashboards/public';
import { DataSourceManagementPluginSetup } from 'src/plugins/data_source_management/public';
// @ts-ignore
import Plotly from 'plotly.js-dist';
import { useHistory, useLocation } from 'react-router-dom';
import React, { useContext, useEffect, useState } from 'react';
import {
  EuiButton,
  EuiCodeBlock,
  EuiFlexGrid,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
  EuiText,
  EuiTitle,
  EuiIconTip,
} from '@elastic/eui';
import { DataSourceContext, QUERY_INSIGHTS } from '../TopNQueries/TopNQueries';
import { QueryGroupAggregateSummary } from './Components/QueryGroupAggregateSummary';
import { QueryGroupSampleQuerySummary } from './Components/QueryGroupSampleQuerySummary';

import { QueryInsightsDashboardsPluginStartDependencies } from '../../types';
import { PageHeader } from '../../components/PageHeader';
import { SearchQueryRecord } from '../../../types/types';
import { retrieveQueryById } from '../Utils/QueryUtils';
import {
  getDataSourceFromUrl,
  QueryInsightsDataSourceMenu,
} from '../../components/DataSourcePicker';

export const QueryGroupDetails = ({
  core,
  depsStart,
  params,
  dataSourceManagement,
}: {
  core: CoreStart;
  params: AppMountParameters;
  dataSourceManagement?: DataSourceManagementPluginSetup;
  depsStart: QueryInsightsDashboardsPluginStartDependencies;
}) => {
  // Get url parameters
  const location = useLocation();
  const searchParams = new URLSearchParams(location.search);
  const id = searchParams.get('id');
  const from = searchParams.get('from');
  const to = searchParams.get('to');

  const [query, setQuery] = useState<SearchQueryRecord | null>(null);
  const { dataSource, setDataSource } = useContext(DataSourceContext)!;

  const convertTime = (unixTime: number) => {
    const date = new Date(unixTime);
    const loc = date.toDateString().split(' ');
    return loc[1] + ' ' + loc[2] + ', ' + loc[3] + ' @ ' + date.toLocaleTimeString('en-US');
  };

  const history = useHistory();

  const fetchQueryDetails = async () => {
    const retrievedQuery = await retrieveQueryById(core, getDataSourceFromUrl().id, from, to, id);
    setQuery(retrievedQuery);
  };

  useEffect(() => {
    if (id && from && to) {
      fetchQueryDetails();
    }
  }, [id, from, to]);

  useEffect(() => {
    if (query) {
      core.chrome.setBreadcrumbs([
        {
          text: 'Query insights',
          href: QUERY_INSIGHTS,
          onClick: (e) => {
            e.preventDefault();
            history.push(QUERY_INSIGHTS);
          },
        },
        { text: `Query group details: ${convertTime(query.timestamp)}` },
      ]);
    }
  }, [core.chrome, history, location, query]);

  useEffect(() => {
    if (query && query.phase_latency_map) {
      let x: number[] = Object.values(query.phase_latency_map);
      if (x.length < 3) {
        x = [0, 0, 0];
      }
      const data = [
        {
          x: x.reverse(),
          y: ['Fetch    ', 'Query    ', 'Expand    '],
          type: 'bar',
          orientation: 'h',
          width: 0.5,
          marker: { color: ['#F990C0', '#1BA9F5', '#7DE2D1'] },
          base: [x[2] + x[1], x[2], 0],
          text: x.map((value) => `${value}ms`),
          textposition: 'outside',
          cliponaxis: false,
        },
      ];
      const layout = {
        autosize: true,
        margin: { l: 80, r: 80, t: 25, b: 15, pad: 0 },
        autorange: true,
        height: 120,
        xaxis: {
          side: 'top',
          zeroline: false,
          ticksuffix: 'ms',
          autorangeoptions: { clipmin: 0 },
          tickfont: { color: '#535966' },
          linecolor: '#D4DAE5',
          gridcolor: '#D4DAE5',
        },
        yaxis: { linecolor: '#D4DAE5' },
      };
      const config = { responsive: true };
      Plotly.newPlot('latency', data, layout, config);
    }
  }, [query]);

  const queryString = query
    ? JSON.stringify(JSON.parse(JSON.stringify(query.source)), null, 2)
    : '';
  const queryDisplay = `{\n  "query": ${queryString ? queryString.replace(/\n/g, '\n  ') : ''}\n}`;

  return (
    <div>
      <PageHeader
        coreStart={core}
        depsStart={depsStart}
        fallBackComponent={
          <>
            <EuiFlexGrid columns={2}>
              <EuiTitle size="l">
                <h1>Query group details</h1>
              </EuiTitle>
              <EuiIconTip
                content="Details for the query group including aggregate statistics and number of queries in the group"
                position="right"
                type="iInCircle"
                aria-label="Details tooltip"
              />
            </EuiFlexGrid>
            <EuiSpacer size="l" />
          </>
        }
      />
      <QueryInsightsDataSourceMenu
        coreStart={core}
        depsStart={depsStart}
        params={params}
        dataSourceManagement={dataSourceManagement}
        setDataSource={setDataSource}
        selectedDataSource={dataSource}
        onManageDataSource={() => {}}
        onSelectedDataSource={fetchQueryDetails}
        dataSourcePickerReadOnly={true}
      />
      <EuiFlexItem>
        <QueryGroupAggregateSummary query={query} />
      </EuiFlexItem>
      <EuiSpacer size="l" />
      <EuiSpacer size="l" />
      <EuiFlexGrid columns={2}>
        <EuiTitle size="l">
          <h1>Sample query details</h1>
        </EuiTitle>
        <EuiIconTip
          content="Details for a sample query in the query group. This is the first query encountered in the group."
          position="right"
          type="iInCircle"
          aria-label="Details tooltip"
        />
      </EuiFlexGrid>
      <EuiSpacer size="l" />
      <EuiFlexItem>
        <QueryGroupSampleQuerySummary query={query} />
        <EuiSpacer size="m" />
        <EuiFlexGrid columns={2}>
          <EuiFlexItem grow={1}>
            <EuiPanel>
              <EuiFlexGroup alignItems="center" justifyContent="spaceBetween">
                <EuiFlexItem>
                  <EuiText size="xs">
                    <h2>Query</h2>
                  </EuiText>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiButton
                    iconSide="right"
                    iconType="popout"
                    target="_blank"
                    href="https://playground.opensearch.org/app/searchRelevance#/"
                  >
                    Open in search comparision
                  </EuiButton>
                </EuiFlexItem>
              </EuiFlexGroup>
              <EuiHorizontalRule margin="xs" />
              <EuiSpacer size="xs" />
              <EuiCodeBlock
                language="jsx"
                paddingSize="m"
                fontSize="s"
                overflowHeight={600}
                isCopyable
              >
                {queryDisplay}
              </EuiCodeBlock>
            </EuiPanel>
          </EuiFlexItem>
          <EuiFlexItem grow={1} style={{ alignSelf: 'start' }}>
            <EuiPanel>
              <EuiText size="xs">
                <h2>Latency</h2>
              </EuiText>
              <EuiHorizontalRule margin="m" />
              <div id="latency" />
            </EuiPanel>
          </EuiFlexItem>
        </EuiFlexGrid>
      </EuiFlexItem>
    </div>
  );
};
