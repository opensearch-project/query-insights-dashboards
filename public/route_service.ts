/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { HttpStart } from '../../../src/core/public';

export class RouteService {
  private http: HttpStart;

  constructor(http: HttpStart) {
    this.http = http;
  }

  async getLocalClusterVersion(): Promise<string | undefined> {
    try {
      const response = await this.http.get('/api/cluster/version');
      return response?.version;
    } catch (error) {
      console.error('Error getting local cluster version:', error);
      return undefined;
    }
  }
}
