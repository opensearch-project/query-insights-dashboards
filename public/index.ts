/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './index.scss';

import { QueryInsightsDashboardsPlugin } from './plugin';

// This exports static code and TypeScript types,
// as well as, OpenSearch Dashboards Platform `plugin()` initializer.
export function plugin() {
  return new QueryInsightsDashboardsPlugin();
}
export { QueryInsightsDashboardsPluginSetup, QueryInsightsDashboardsPluginStart } from './types';
