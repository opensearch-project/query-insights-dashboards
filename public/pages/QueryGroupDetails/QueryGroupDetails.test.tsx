/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router-dom';
import { QueryGroupDetails } from './QueryGroupDetails';
import { CoreStart } from 'opensearch-dashboards/public';
import React from 'react';
import { mockQueries } from '../../../test/mocks/mockQueries';
import '@testing-library/jest-dom/extend-expect';
import { retrieveQueryById } from '../../../common/utils/QueryUtils';
import { DataSourceContext } from '../TopNQueries/TopNQueries';

jest.mock('object-hash', () => jest.fn(() => '8c1e50c035663459d567fa11d8eb494d'));

jest.mock('plotly.js-dist', () => ({
  newPlot: jest.fn(),
  react: jest.fn(),
  relayout: jest.fn(),
}));

jest.mock('../../../common/utils/QueryUtils', () => ({
  retrieveQueryById: jest.fn(),
}));

jest.mock('react-ace', () => ({
  __esModule: true,
  default: () => <div>Mocked Ace Editor</div>,
}));

const dataSourceMenuMock = jest.fn(() => <div>Mock DataSourceMenu</div>);

const dataSourceManagementMock = {
  ui: {
    getDataSourceMenu: jest.fn().mockReturnValue(dataSourceMenuMock),
  },
};
const mockDataSourceContext = {
  dataSource: { id: 'test', label: 'Test' },
  setDataSource: jest.fn(),
};

const mockQuery = mockQueries[0];

describe('QueryGroupDetails', () => {
  const coreMock = ({
    chrome: {
      setBreadcrumbs: jest.fn(),
    },
    uiSettings: {
      get: jest.fn().mockReturnValue(false),
    },
  } as unknown) as CoreStart;

  beforeEach(() => {
    jest.clearAllMocks();
    (retrieveQueryById as jest.Mock).mockResolvedValue(mockQuery);
  });

  const renderComponent = () => {
    return render(
      <MemoryRouter
        initialEntries={[
          '/query-group-details?id=mockId&from=1632441600000&to=1632528000000&verbose=true',
        ]}
      >
        <DataSourceContext.Provider value={mockDataSourceContext}>
          <Route path="/query-group-details">
            <QueryGroupDetails
              core={coreMock}
              depsStart={{ navigation: {} }}
              params={{} as any}
              dataSourceManagement={dataSourceManagementMock}
            />
          </Route>
        </DataSourceContext.Provider>
      </MemoryRouter>
    );
  };

  it('renders the QueryGroupDetails component', async () => {
    renderComponent();

    expect(screen.getByText('Query group details')).toBeInTheDocument();
    expect(screen.getByText('Sample query details')).toBeInTheDocument();

    await waitFor(() => {
      expect(coreMock.chrome.setBreadcrumbs).toHaveBeenCalledWith([
        expect.objectContaining({ text: 'Query insights' }),
        expect.objectContaining({ text: expect.stringMatching(/^Query group details: .+/) }),
      ]);
    });
  });

  it('fetches and displays query group data', async () => {
    renderComponent();

    await waitFor(() => {
      expect(retrieveQueryById).toHaveBeenCalledWith(
        coreMock,
        undefined,
        '1632441600000',
        '1632528000000',
        'mockId',
        true
      );
    });

    expect(screen.getByText('Query')).toBeInTheDocument();
    expect(screen.getByText('Latency')).toBeInTheDocument();
  });

  it('renders latency bar chart', async () => {
    renderComponent();

    await waitFor(() => {
      expect(screen.getAllByText('Latency')).toHaveLength(1);
    });

    expect(document.getElementById('latency')).toBeInTheDocument();
  });

  it('displays query details', async () => {
    const { container } = renderComponent();

    await waitFor(() => {
      expect(screen.getByText('Open in search comparision')).toBeInTheDocument();
    });

    const codeBlock = container.querySelector('.euiCodeBlock__pre');
    expect(codeBlock).toBeInTheDocument();

    expect(codeBlock?.textContent).toContain('"query"');
    expect(codeBlock?.textContent).toMatch(/{\s*"query":/);
  });

  it('renders tooltips', () => {
    renderComponent();

    const tooltips = screen.getAllByLabelText('Details tooltip');
    expect(tooltips).toHaveLength(2);
  });

  it('renders correct breadcrumb based on query timestamp', async () => {
    jest.spyOn(Date.prototype, 'toDateString').mockReturnValue('Mon Sep 24 2021');
    jest.spyOn(Date.prototype, 'toLocaleTimeString').mockReturnValue('12:00:00 AM');

    renderComponent();

    await waitFor(() => {
      expect(coreMock.chrome.setBreadcrumbs).toHaveBeenCalledWith([
        {
          text: 'Query insights',
          href: '/queryInsights',
          onClick: expect.any(Function),
        },
        {
          text: 'Query group details: Sep 24, 2021 @ 12:00:00 AM',
        },
      ]);
    });
  });

  it('matches snapshot', async () => {
    const { container } = renderComponent();

    await waitFor(() => {
      expect(retrieveQueryById).toHaveBeenCalled();
    });

    const dateElements = container.getElementsByClassName('euiText euiText--extraSmall');
    Array.from(dateElements).forEach((element) => {
      if (element.textContent?.includes('@')) {
        element.textContent = 'Sep 24, 2021 @ 12:00:00 AM';
      }
    });

    expect(container).toMatchSnapshot();
  });
});
