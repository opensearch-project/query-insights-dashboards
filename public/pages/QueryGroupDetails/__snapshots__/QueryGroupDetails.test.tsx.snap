// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryGroupDetails matches snapshot 1`] = `
<div>
  <div>
    <div
      class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--halves euiFlexGrid--responsive"
    >
      <h1
        class="euiTitle euiTitle--large"
      >
        Query group details
      </h1>
      <span
        class="euiToolTipAnchor"
      >
        <svg
          aria-label="Details tooltip"
          class="euiIcon euiIcon--medium"
          focusable="true"
          height="16"
          role="img"
          tabindex="0"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.5 11.508 7.468 8H6.25V7h2.401l.03 3.508H9.8v1H7.5Zm-.25-6.202a.83.83 0 0 1 .207-.577c.137-.153.334-.229.59-.229.256 0 .454.076.594.23.14.152.209.345.209.576 0 .228-.07.417-.21.568-.14.15-.337.226-.593.226-.256 0-.453-.075-.59-.226a.81.81 0 0 1-.207-.568ZM8 13A5 5 0 1 0 8 3a5 5 0 0 0 0 10Zm0 1A6 6 0 1 1 8 2a6 6 0 0 1 0 12Z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiFlexItem"
    >
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
      >
        <div
          class="euiText euiText--extraSmall"
        >
          <h2>
            Aggregate summary for 
            8
             
            queries
          </h2>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginMedium"
        />
        <div
          class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--fourths euiFlexGrid--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Id
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              8c1e50c035663459d567fa11d8eb494d
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Average Latency
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              2.50 ms
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Average CPU Time
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              1.42 ms
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Average Memory Usage
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              16528.00 B
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Group by
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              SIMILARITY
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--halves euiFlexGrid--responsive"
    >
      <h1
        class="euiTitle euiTitle--large"
      >
        Sample query details
      </h1>
      <span
        class="euiToolTipAnchor"
      >
        <svg
          aria-label="Details tooltip"
          class="euiIcon euiIcon--medium"
          focusable="true"
          height="16"
          role="img"
          tabindex="0"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.5 11.508 7.468 8H6.25V7h2.401l.03 3.508H9.8v1H7.5Zm-.25-6.202a.83.83 0 0 1 .207-.577c.137-.153.334-.229.59-.229.256 0 .454.076.594.23.14.152.209.345.209.576 0 .228-.07.417-.21.568-.14.15-.337.226-.593.226-.256 0-.453-.075-.59-.226a.81.81 0 0 1-.207-.568ZM8 13A5 5 0 1 0 8 3a5 5 0 0 0 0 10Zm0 1A6 6 0 1 1 8 2a6 6 0 0 1 0 12Z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiFlexItem"
    >
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
      >
        <div
          class="euiText euiText--extraSmall"
        >
          <h2>
            Sample query summary
          </h2>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginMedium"
        />
        <div
          class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--fourths euiFlexGrid--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Timestamp
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              Sep 24, 2021 @ 12:00:00 AM
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Indices
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              my-index
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Search Type
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              query then fetch
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Coordinator Node ID
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              HjvgxQ4AQTiddd43OV7pJA
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Total Shards
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              1
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--m"
      />
      <div
        class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--halves euiFlexGrid--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrow1"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              >
                <div
                  class="euiText euiText--extraSmall"
                >
                  <h2>
                    Query
                  </h2>
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <a
                  class="euiButton euiButton--primary"
                  href="https://playground.opensearch.org/app/searchRelevance#/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  <span
                    class="euiButtonContent euiButtonContent--iconRight euiButton__content"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon--inherit euiButtonContent__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M13 8.5a.5.5 0 1 1 1 0V12a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h3.5a.5.5 0 0 1 0 1H4a1 1 0 0 0-1 1v8a1 1 0 0 0 1 1h8a1 1 0 0 0 1-1V8.5Zm-5.12.339a.5.5 0 1 1-.706-.707L13.305 2H10.5a.5.5 0 1 1 0-1H14a1 1 0 0 1 1 1v3.5a.5.5 0 1 1-1 0V2.72L7.88 8.838Z"
                      />
                    </svg>
                    <span
                      class="euiButton__text"
                    >
                      Open in search comparision
                    </span>
                  </span>
                </a>
              </div>
            </div>
            <hr
              class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
            />
            <div
              class="euiSpacer euiSpacer--xs"
            />
            <div
              class="euiCodeBlock euiCodeBlock--fontSmall euiCodeBlock--paddingMedium euiCodeBlock--hasControls prismjs language-jsx"
              style="max-height: 600px;"
            >
              <pre
                class="euiCodeBlock__pre euiCodeBlock__pre--whiteSpacePreWrap"
                style="max-height: 600px;"
                tabindex="-1"
              >
                <code
                  class="euiCodeBlock__code jsx"
                >
                  <span
                    class="euiCodeBlock__line"
                  >
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                      
                    <span
                      class="token string-property property"
                    >
                      "query"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                        
                    <span
                      class="token string-property property"
                    >
                      "size"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token number"
                    >
                      0
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                        
                    <span
                      class="token string-property property"
                    >
                      "aggregations"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                          
                    <span
                      class="token string-property property"
                    >
                      "average_age"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                            
                    <span
                      class="token string-property property"
                    >
                      "avg"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                              
                    <span
                      class="token string-property property"
                    >
                      "field"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token string"
                    >
                      "age"
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                            
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                          
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                        
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                      
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                    
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                  </span>
                </code>
              </pre>
              <div
                class="euiCodeBlock__controls"
              >
                <button
                  aria-label="Expand"
                  class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiCodeBlock__fullScreenButton"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon--inherit euiButtonIcon__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M13 3v4h-1V4H9V3h4ZM3 3h4v1H4v3H3V3Zm10 10H9v-1h3V9h1v4ZM3 13V9h1v3h3v1H3ZM0 1.994C0 .893.895 0 1.994 0h12.012C15.107 0 16 .895 16 1.994v12.012A1.995 1.995 0 0 1 14.006 16H1.994A1.995 1.995 0 0 1 0 14.006V1.994Zm1 0v12.012c0 .548.446.994.994.994h12.012a.995.995 0 0 0 .994-.994V1.994A.995.995 0 0 0 14.006 1H1.994A.995.995 0 0 0 1 1.994Z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
                <div
                  class="euiCodeBlock__copyButton"
                >
                  <span
                    class="euiToolTipAnchor"
                  >
                    <button
                      aria-label="Copy"
                      class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiIcon--inherit euiButtonIcon__icon"
                        focusable="false"
                        height="16"
                        role="img"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M11.4 0c.235 0 .46.099.622.273l2.743 3c.151.162.235.378.235.602v9.25a.867.867 0 0 1-.857.875H3.857A.867.867 0 0 1 3 13.125V.875C3 .392 3.384 0 3.857 0H11.4ZM14 4h-2.6a.4.4 0 0 1-.4-.4V1H4v12h10V4Z"
                        />
                        <path
                          d="M3 1H2a1 1 0 0 0-1 1v13a1 1 0 0 0 1 1h10a1 1 0 0 0 1-1v-1h-1v1H2V2h1V1Z"
                        />
                      </svg>
                    </button>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrow1"
          style="align-self: start;"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h2>
                Latency
              </h2>
            </div>
            <hr
              class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginMedium"
            />
            <div
              id="latency"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
